src/modelframe.jl

```diff
@@ -1,261 +1,102 @@
-"""
-function StatsBase.model_response(mf::ModelFrame)
-    if mf.terms.response
-        _response = mf.df[mf.terms.eterms[1]]
-        T = Missings.T(eltype(_response))
-        convert(Array{T}, _response)
-    else
-        error("Model formula one-sided")
+setcontrasts!(mf::ModelFrame; kwargs...) = setcontrasts!(mf, Dict(kwargs))
+function setcontrasts!(mf::ModelFrame, contrasts::Dict{Symbol})
+    # TODO: don't consume the whole table again if it's not needed
+    new_schema = schema([term(k) for k in keys(contrasts)], mf.data, contrasts)
+
+    # warn of no-op for keys that dont correspond to known terms from old schema
+    unknown_keys = [k for k in keys(new_schema) if !haskey(mf.schema, k)]
+    if !isempty(unknown_keys)
+        unknown_keys_str = join(unknown_keys, ", ", " and ")
+        @warn "setcontrasts! for terms " * unknown_keys_str *
+            " has no effect since they are not found in original schema"
     end
-end
 
-
-"""
-    termnames(term::Symbol, col)
-Returns a vector of strings with the names of the coefficients
-associated with a term.  If the column corresponding to the term
-is not categorical, a one-element vector is returned.
-"""
-termnames(term::Symbol, col) = [string(term)]
-function termnames(term::Symbol, mf::ModelFrame; non_redundant::Bool = false)
-    if haskey(mf.contrasts, term)
-        termnames(term, mf.df[term],
-                  non_redundant ?
-                  convert(ContrastsMatrix{FullDummyCoding}, mf.contrasts[term]) :
-                  mf.contrasts[term])
-    else
-        termnames(term, mf.df[term])
-    end
+    # apply only the re-mapped terms
+    mf.f = apply_schema(mf.f, new_schema, mf.model)
+    merge!(mf.schema, new_schema)
+    mf
 end
 
-termnames(term::Symbol, col::Any, contrast::ContrastsMatrix) =
-    ["$term: $name" for name in contrast.termnames]
-
 
-function expandtermnames(term::Vector)
-    if length(term) == 1
-        return term[1]
-    else
-        return foldr((a,b) -> vec([string(lev1, " & ", lev2) for
-                                   lev1 in a,
-                                   lev2 in b]),
-                     term)
-    end
-end
 
+const AbstractFloatMatrix{T<:AbstractFloat} =  AbstractMatrix{T}
 
-"""
-    coefnames(mf::ModelFrame)
-Returns a vector of coefficient names constructed from the Terms
-member and the types of the evaluation columns.
-"""
-function StatsBase.coefnames(mf::ModelFrame)
-    terms = droprandomeffects(dropresponse!(mf.terms))
-
-    ## strategy mirrors ModelMatrx constructor:
-    eterm_names = Dict{Tuple{Symbol,Bool}, Vector{String}}()
-    term_names = Vector{Vector{String}}()
-
-    if terms.intercept
-        push!(term_names, String["(Intercept)"])
-    end
-
-    factors = terms.factors
-
-    for (i_term, term) in enumerate(terms.terms)
-
-        ## names for columns for eval terms
-        names = Vector{Vector{String}}()
+mutable struct ModelMatrix{T <: AbstractFloatMatrix}
```

Comment:
As noted before you can now replace this with `T<:AbstractMatrix{<:AbstractFloat}` and remove the `AbstractFloatMatrix` alias.